docs/Recovery-Postcard.md

```diff
@@ -0,0 +1,143 @@
+
+
+### Printing order
+
+1. PAS and RPPS must have pre-shared asymmetric EC keys (P256r1), leading to 32bytes long shared secret (ECDH). Let's call this as `RPPS_MASTER_SHARED_SECRET`.
+
+1. RPOS asks PAS for postcard creation for one particular user
+
+1. PAS will calculate
+   ```java
+   byte[] SHARED_SECRET
+   if (RPPS_MASTER_SHARED_SECRET == null) {
+       // Mode useful for RC associated with activation
+       SHARED_SECRET = Generator.randomBytes(32)
+   } else {
+       // Mode useful for RC postcard creation
+       SHARED_SECRET = RPPS_MASTER_SHARED_SECRET
+   }
+   
+   repeat 
+   {
+       byte[] NONCE = Generator.randomBytes(32)
+       byte[] SHARED_POSTCARD = KDF_X9_63_SHA256.derive(SHARED_SECRET, NONCE)
+       byte[] RECOVERY_CODE_BYTES = ByteUtils.subarray(SHARED_POSTCARD, 0, 10)
+       String RECOVERY_CODE = Generator.buildActivationCode(RECOVERY_CODE_BYTES)
+       if RECOVERY_CODE doesn't exist in DB {
+           break
+       }
+   }
+
+   byte[] RECOVERY_PUK_BASE_KEY = ByteUtils.subarray(SHARED_POSTCARD, 10, 16)
+
+   // Prepare array of PUK derivation indexes, set to 0
+   long[] PUK_DERIVATION_INDEX_ARRAY[PUK_COUNT] = { 0 }
+   
+   // Prepare array of PUK values
+   String[] PUK_ARRAY[PUK_COUNT] = { null }
+
+   for I = 0; I < PUK_COUNT; i++ 
+   {
+       repeat
+       {
+           long PUK_DERIVATION_INDEX = ByteUtils.getLong(Generator.randomBytes(8))
+           byte[] PUK_KEY = KDF.derive(RECOVERY_PUK_BASE_KEY, PUK_DERIVATION_INDEX)
+           long PUK_VALUE = (ByteUtils.getLong(ByteUtils.truncate(PUK_KEY, 8)) & 0xFFFFFFFFFF) % Math.pow(10,10)
+           String PUK = String.format("%010d", PUK_VALUE)
+           if  !PUK_ARRAY.contains(PUK) {
+               PUK_DERIVATION_INDEX_ARRAY[i] = PUK_DERIVATION_INDEX
+               PUK_ARRAY[i] = PUK
+               break
+           }
+       }   
+   }
+   ```
+   
+1. PAS store following values to the database:
+    - `RECOVERY_CODE`
+    - For each value in `PUK_ARRAY` stores argon2 hashed value
+
+1. PAS will return following values to RPOS
+    - `NONCE`
+    - `PUK_DERIVATION_INDEX_ARRAY`
+    
+1. RPOS will construct printing request for RPPS:
+   ```json
+   {
+      "bankClient": {
+          "gender": "M",
+          "fullName": "Franta Novák",
+          "company": "Wultra s.r.o.",
+          "streetName": "Budějovická",
+          "streetNumber": "779/3a",
+          "city": "Praha 4",
+          "zip": "14000",
+          "country": "CZ"
+      },
+      "postcard": {
+          "identifier": "SOME-ISSUE-IDENTIFIER",
+          "nonce": "32bytes, encoded in BASE64",
+          "pukDerivationIndexes": [
+                323213,
+                123123,
+                535,
+                31329854,
+                432432
+          ]
+      }
+   }
+   ```
+   _Note that information above is just an example and the actual payload may be slightly different._
```

Comment:
"Note that the JSON structure above is just an example to illustrate how nonce and PUK derivation indexes may be passed with a permanent address of a recovery postcard recipient. The actual payload format may be slightly different for different business purposes."